### ISSUE TITLE
Configurable Token Cache

### ISSUE THREAD

**Issue by kamaz** [commercetools/nodejs#1497]
### Description

As part **sdk-middleware-auth** you cache `auth` and reuse when it is present but that only supports a single client. For details: https://github.com/commercetools/nodejs/blob/master/packages/sdk-middleware-auth/src/base-auth-flow.js#L68

As part of our implementation, we have to support multiple clients and we would like to extend **sdk-middleware-auth** to add that support. 

The proposal for the change: 
1. Extend `AuthMiddlewareOptions` to allow passing custom `tokenCache` implementation
2. Change interface for `tokenCache` (https://github.com/commercetools/nodejs/blob/master/types/sdk.js#L185) to: 

```js
type TokenCache {
  get: (client_id?: string) => TokenStore
  set: (cache: TokenStore, client_id?: string) => TokenStore
}
```

For backward compatibility keep the order of parameters and make the`client_id` parameter optional.

3. Change the code occurrences to pass the `client_id` to all gets and sets 

Please let me know if you would be interested with above I would be happy to contribute.

**Comment by emmenko** [comment]
That would be nice yes. We should also do the same for the `TokenProvider` (https://github.com/commercetools/nodejs/blob/master/packages/sdk-auth/src/tokenProvider.js) I guess.

If you feel like it, we are happy to welcome a PR from you 🤗

**Comment by kamaz** [comment]
I've created #1498 

I had a look at `TokenProvider` and that looks like a wrapper around `SdkAuth`. IMHO making it `client` specific may not be a good idea. 

Please have a look at the PR and let me know what you think? 

Also in the context of the change it would be good to hear how would you see that working with `TokenProvider`.

**Comment by daern91** [comment]
Fixed in #1498, thanks again @kamaz! 💯🏅

**Comment by studentIvan** [comment]
Configurable Token Cache is good, but the big part of the data providers, even files are async and it does not support the async token management. For example cloudflare workers have the KV Storage and I would like to have something like this

```ts
function cfTokenStore() {
  return {
    get: () => {
      return JSON.parse(await BFF_KV.get('tokendata'));
    },
    set: (val: { token: string; expirationTime: number }) => {
      /** make async set on set, with the value work as default */
      await BFF_KV.put('tokendata', JSON.stringify(val));
      return val;
    }
  };
}
```
